File: src/compiler/z42.Tests/TokensTests.cs
using FluentAssertions;
using Z42.IR;

namespace Z42.Tests;

/// <summary>
/// Phase 3 (tokenize-ir-and-zbc-bump, 2026-05-09): mirrors Rust
/// <c>runtime/src/metadata/tokens_tests.rs</c>. Verifies token space
/// layout and IsImport / ImportIdx semantics across all 4 importable kinds.
/// </summary>
public class TokensTests
{
    [Fact]
    public void TokenConstants_HaveExpectedValues()
    {
        TokenConsts.Unresolved.Should().Be(0xFFFF_FFFFu);
        TokenConsts.ImportBase.Should().Be(0x8000_0000u);
    }

    [Fact]
    public void UnresolvedToken_IsNotResolved()
    {
        MethodId.Unresolved.IsResolved.Should().BeFalse();
        TypeId.Unresolved.IsResolved.Should().BeFalse();
        StaticFieldId.Unresolved.IsResolved.Should().BeFalse();
        BuiltinId.Unresolved.IsResolved.Should().BeFalse();
    }

    [Fact]
    public void IntraModuleToken_IsResolvedAndNotImport()
    {
        var m = new MethodId(0);
        m.IsResolved.Should().BeTrue();
        m.IsImport.Should().BeFalse();

        var t = new TypeId(42);
        t.IsResolved.Should().BeTrue();
        t.IsImport.Should().BeFalse();

        var s = new StaticFieldId(0x7FFF_FFFFu);
        s.IsResolved.Should().BeTrue();
        s.IsImport.Should().BeFalse();
    }

    [Fact]
    public void ImportToken_IsResolvedAndImport()
    {
        var m = new MethodId(TokenConsts.ImportBase);
        m.IsResolved.Should().BeTrue();
        m.IsImport.Should().BeTrue();
        m.ImportIdx.Should().Be(0u);

        var t = new TypeId(TokenConsts.ImportBase + 5u);
        t.IsImport.Should().BeTrue();
        t.ImportIdx.Should().Be(5u);

        var s = new StaticFieldId(TokenConsts.ImportBase + 0x1234u);
        s.IsImport.Should().BeTrue();
        s.ImportIdx.Should().Be(0x1234u);
    }

    [Fact]
    public void UnresolvedToken_IsNotImport()
    {
        // UNRESOLVED is in the >= ImportBase range numerically but specifically
        // excluded from import semantics so callers can tell the two apart.
        MethodId.Unresolved.IsImport.Should().BeFalse();
        TypeId.Unresolved.IsImport.Should().BeFalse();
    }

    [Fact]
    public void TokenTypes_AreDistinct()
    {
        // Compile-time check — these would not compile if MethodId == TypeId
        // (record struct types are distinct nominal types).
        var m = new MethodId(1);
        var t = new TypeId(1);
        m.Value.Should().Be(t.Value); // same underlying u32 …
        // … but you cannot assign one to the other:
        // MethodId bad = t;  // does not compile.
    }

    [Fact]
    public void ImportKind_HasFourKinds()
    {
        ((byte)ImportKind.Method).Should().Be(0x01);
        ((byte)ImportKind.Type).Should().Be(0x02);
        ((byte)ImportKind.StaticField).Should().Be(0x03);
        ((byte)ImportKind.Builtin).Should().Be(0x04);
    }

    [Fact]
    public void ToString_ReturnsHumanReadable()
    {
        MethodId.Unresolved.ToString().Should().Be("MethodId(unresolved)");
        new MethodId(7).ToString().Should().Be("MethodId(7)");
        new MethodId(TokenConsts.ImportBase + 3).ToString().Should().Be("MethodId(import 3)");
        new TypeId(0).ToString().Should().Be("TypeId(0)");
        new StaticFieldId(99).ToString().Should().Be("StaticFieldId(99)");
        new BuiltinId(2).ToString().Should().Be("BuiltinId(2)");
    }
}

File: src/runtime/src/metadata/tokens_tests.rs
use super::*;

#[test]
fn unresolved_sentinel_is_u32_max() {
    assert_eq!(UNRESOLVED, u32::MAX);
    assert!(!MethodId::UNRESOLVED.is_resolved());
    assert!(!TypeId::UNRESOLVED.is_resolved());
    assert!(!BuiltinId::UNRESOLVED.is_resolved());
    assert!(!FieldId::UNRESOLVED.is_resolved());
    assert!(!StaticFieldId::UNRESOLVED.is_resolved());
    assert!(!VTableSlot::UNRESOLVED.is_resolved());
}

#[test]
fn resolved_token_reports_resolved() {
    assert!(MethodId(0).is_resolved());
    assert!(MethodId(42).is_resolved());
    assert!(BuiltinId(7).is_resolved());
}

#[test]
fn token_types_are_distinct() {
    // Compile-time check — these would fail to compile if MethodId == TypeId.
    let _m: MethodId = MethodId(1);
    let _t: TypeId = TypeId(1);
    // Cannot mix: `let _: MethodId = TypeId(1);` would not compile.
}

// ── Phase 3 IMPORT_BASE semantics ────────────────────────────────────────

#[test]
fn import_base_constant_is_high_bit() {
    assert_eq!(IMPORT_BASE, 0x8000_0000);
}

#[test]
fn intra_module_token_is_not_import() {
    assert!(!MethodId(0).is_import());
    assert!(!MethodId(42).is_import());
    assert!(!MethodId(0x7FFF_FFFF).is_import());
    assert!(!TypeId(0).is_import());
    assert!(!StaticFieldId(100).is_import());
}

#[test]
fn import_token_reports_import() {
    let m = MethodId(IMPORT_BASE);
    assert!(m.is_import());
    assert_eq!(m.import_idx(), 0);

    let t = TypeId(IMPORT_BASE + 5);
    assert!(t.is_import());
    assert_eq!(t.import_idx(), 5);

    let s = StaticFieldId(IMPORT_BASE + 0x1234);
    assert!(s.is_import());
    assert_eq!(s.import_idx(), 0x1234);
}

#[test]
fn unresolved_is_neither_resolved_nor_import() {
    let u = MethodId::UNRESOLVED;
    assert!(!u.is_resolved());
    assert!(!u.is_import()); // UNRESOLVED specifically excluded from import range
}

#[test]
fn import_token_is_resolved() {
    // An imported token IS resolved (just into import_table, not module.functions);
    // the dispatch site uses is_import() to choose the path.
    assert!(MethodId(IMPORT_BASE).is_resolved());
    assert!(MethodId(IMPORT_BASE + 99).is_resolved());
}

#[test]
#[should_panic(expected = "import_idx() on non-import token")]
fn import_idx_on_intra_module_panics_in_debug() {
    let m = MethodId(0);
    let _ = m.import_idx();
}
